doc/ml_features.rst

```diff
@@ -0,0 +1,337 @@
+``serialize_model``.
+
+The method ``freeze_model`` is thought to be used in conjunction
+with SmartRedis ``set_model_from_file``. The following is a typical
+workflow, first we define the model:
+
+.. code-block:: python
+
+    import numpy as np
+    from smartredis import Client
+    from tensorflow import keras
+    from smartsim.ml.tf import freeze_model
+
+    model = keras.Sequential(
+        layers=[
+            keras.layers.InputLayer(input_shape=(28, 28), name="input"),
+            keras.layers.Flatten(input_shape=(28, 28), name="flatten"),
+            keras.layers.Dense(128, activation="relu", name="dense"),
+            keras.layers.Dense(10, activation="softmax", name="output"),
+        ],
+        name="FCN",
+    )
+
+    # Compile model with optimizer
+    model.compile(
+        optimizer="adam", loss="sparse_categorical_crossentropy", metrics=["accuracy"]
+    )
+
+The model is now ready to be serialized and stored on the filesystem:
+
+.. code-block:: python
+
+    model_path, inputs, outputs = freeze_model(model, '.', "mnist.pb")
+
+Note that ``freeze_model`` conveniently returns the path to the serialized model file,
+and the names of the input and output layers, which are needed to upload the TensorFlow
+model on the DB, as shown in the following code snippet, where we also upload a
+synthetic sample to be passed to the model. Notice that we could also upload a batch
+of samples, instead of a single one.
+
+.. code-block:: python
+
+    client = Client(cluster=False)
+    model_key = "tf_mnist"
+    client.set_model_from_file(
+        model_key, model_path, "TF", device="GPU", inputs=inputs, outputs=outputs
+    )
+    mnist_image = np.random.rand(1, 28, 28).astype(np.float32)
+    client.put_tensor("mnist_input", mnist_image)
+
+For details about ``set_model_from_file``, please
+refer to :ref:`SmartRedis API <smartredis-api>`.
+Finally, the model can be run on the sample and the output is ready to be downloaded.
+
+.. code-block:: python
+
+    client.run_model(model_key, "mnist_input", "mnist_output")
+    pred = client.get_tensor("mnist_output")
+
+
+If storing the model as a file is not needed, then it can just be kept in memory
+after serialization, using ``serialize_model`` after compiling the model. The same
+workflow we saw in the previous example can then basically be achieved by replacing
+``set_model_from_file`` with ``set_model``:
+
+.. code-block:: python
+
+    # ... standard imports
+    from smartsim.ml.tf import serialize_model
+
+    # ... define, instantiate, and compile Keras model
+
+    serialized_model, inputs, outputs = serialize_model(model)
+
+    client = Client(cluster=False)
+    model_key = "tf_mnist_serialized"
+    client.set_model(
+        model_key, serialized_model, "TF", device="GPU", inputs=inputs, outputs=outputs
+    )
+    mnist_image = np.random.rand(1, 28, 28).astype(np.float32)
+    client.put_tensor("mnist_input", mnist_image)
```

Comment:
comment the code so i know when you call put_tensor youre putting the db tensors into the model

Reply:
or scuse me "throwing the tensors up in the db so u can pass them to the model below"

Reply:
do the same for the other examples [: